### PR TITLE
add tox env for minimal required dependencies

### PR DESCRIPTION
**Pull request by jkowalleck**
setup to test lowest possible requirements, while still being able to test the locked/pinned dev-dependencies

caused by #44

### CODE REVIEW

**Comment by jkowalleck** [comment]
@madpah some of the types are found to be errorneous in py36 - later versions are fine with them.
could you see if you can fix them?

File: .github/workflows/poetry.yml [hunk 1]
```diff
@@ -14,10 +14,12 @@ on:
 
 env:
   REPORTS_DIR: CI_reports
+  PYTHON_VERISON_DEFAULT: "3.10"
+  POETRY_VERSION: "1.1.11"
 
 jobs:
   coding-standards:
-    name: Linting & Coding Standards
+    name: Linting & CodingStandards
```
Comment (jkowalleck) [inline .github/workflows/poetry.yml:22]: shortened some labels, so they fit on the screen

File: .github/workflows/poetry.yml [hunk 2]
```diff
@@ -27,25 +29,35 @@ jobs:
...
-          poetry-version: 1.1.8
+          poetry-version: ${{ env.POETRY_VERSION }}
       - uses: actions/cache@v2
         with:
           path: ~/.cache/pypoetry/virtualenvs
-          key: ${{ runner.os }}-poetry-${{ hashFiles('poetry.lock') }}
+          key: ${{ runner.os }}-${{ env.PYTHON_VERISON_DEFAULT }}-poetry${{ env.POETRY_VERSION }}-${{ hashFiles('poetry.lock') }}
       - name: Install dependencies
-        run: poetry install
+        run: poetry install --no-root
```
Comment (jkowalleck) [inline .github/workflows/poetry.yml:44]: dont need the package itself installed. it is installed in the tox later, if needed

**Comment by madpah** [comment]
@jkowalleck - this is a cool solution to the problem 👍 

Will take a look at the py36 type errors.

**Comment by madpah** [comment]
@jkowalleck - got rid of a bunch of typing errors for Py 3.6, but the remainder seem to be being incorrectly reported in conditional imports (that only apply to Py >= 3.8).

Thoughts?

**Comment by jkowalleck** [comment]
> @jkowalleck - got rid of a bunch of typing errors for Py 3.6, but the remainder seem to be being incorrectly reported in conditional imports (that only apply to Py >= 3.8).
> 
> Thoughts?

great job.

will look into this later and see if there is any solution that works for us.

**Comment by jkowalleck** [comment]
@madpah i think i got the typing correct for `EnvironmentParser._get_metadata_for_package()`
But now the type checker finds a lot of issues - which i expect to be corect positives.

This would mean that the implementation is wrong for python<3.8
Are there any unittests for `EnvironmentParser.__init__()`?  do they run in py36 properly and actually cover the lines like 
```
if 'Author' in i_metadata.keys()
```
and later?

File: cyclonedx/parser/environment.py [hunk 3]
```diff
@@ -74,8 +72,5 @@ def __init__(self) -> None:
             self._components.append(c)
 
     @staticmethod
-    def _get_metadata_for_package(package_name: str) -> email.message.Message:
-        if sys.version_info >= (3, 8, 0):
-            return metadata(package_name)
```
Comment (jkowalleck) [inline cyclonedx/parser/environment.py:79]: this was the same for py38 and below.
i suspect some internal implementation got lost on the way ...

**Comment by jkowalleck** [comment]
this comments purpose is to move this PR to top of list.

**Comment by madpah** [comment]
Sure @jkowalleck - will check this now.

**Comment by madpah** [comment]
@jkowalleck - looks like I've managed to resolve a bunch of the typing issues in `EnvironmentParser` for Py3.6, but one left:

```
context  [no-untyped-call]
    ...CycloneDX', name='cyclonedx-python-lib', version=meta_version('cyclone...
                                                        ^
Found 1 error in 1 file (checked 19 source files)
ERROR: InvocationError for command /home/runner/work/cyclonedx-python-lib/cyclonedx-python-lib/.tox/mypy-lowest/bin/poetry run mypy (exited with code 1)
```

Would you be happy for us to ignore this or even disable this rule?

**Comment by madpah** [comment]
Although weirdly it's passing in `py3.6-locked`, but not `py3.6-lowest` and the dependency `importlib-metadata` is the same version 😕

File: cyclonedx/model/bom.py [hunk 4]
```diff
@@ -83,10 +83,10 @@ def __repr__(self) -> str:
...
 
 
-if sys.version_info >= (3, 8, 0):
+if sys.version_info >= (3, 8):
     from importlib.metadata import version as meta_version
 else:
-    from importlib_metadata import version as meta_version  # type: ignore
+    from importlib_metadata import version as meta_version
 
 try:
```
Comment (jkowalleck) [inline cyclonedx/model/bom.py:91]: suggested change: 

```diff
- try:
-     ThisTool = Tool(vendor='CycloneDX', name='cyclonedx-python-lib', version=meta_version('cyclonedx-python-lib'))
- except Exception:
-     ThisTool = Tool(vendor='CycloneDX', name='cyclonedx-python-lib', version='UNKNOWN')
+ try:
+    __ThisToolVersion = str(meta_version('cyclonedx-python-lib'))
+ except Exception:
+     __ThisToolVersion = None
+  ThisTool = Tool(vendor='CycloneDX', name='cyclonedx-python-lib', version=__ThisToolVersion or 'UNKNOWN')
```

**Comment by jkowalleck** [comment]
re: https://github.com/CycloneDX/cyclonedx-python-lib/pull/79#issuecomment-984680092

@madpah what if we would take everything that is not expected to raise an error , and move it out of the try/expect? 
would make code clearer and probably get rid of the typing issue - which might happen for a reason.

i am not sure if this solves the issue, but here is a patch: 
https://github.com/CycloneDX/cyclonedx-python-lib/pull/79#discussion_r761282158

could you see if this fixes the issue?

**Comment by jkowalleck** [comment]
I think the tox setup is not great, but it does the job.
Do you think we can roll with it, @madpah ?

**Review by madpah — APPROVED** [review]
All looks good enough to be @jkowalleck

**Pull request merged**